File: 931-minimum-falling-path-sum/931-minimum-falling-path-sum.java
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        int m=matrix.length;
        int n=matrix[0].length;
        
        int[] prev=new int[n];
        
        
        Arrays.fill(prev,0);
        
       
       
       for(int j=0;j<n;j++){
           prev[j]=matrix[m-1][j];
       }
        
        
        for(int i=m-2;i>=0;i--){
            
            int[] curr=new int[n];
            
            for(int j=0;j<n;j++){

                int up=prev[j];

                int left=0;
                if(j-1>=0) left+=prev[j-1];
                else left+=Integer.MAX_VALUE; 
                
              
                int right=0;
                if(j+1<n) right+=prev[j+1];
                else right+=Integer.MAX_VALUE; ;
                
                curr[j]=matrix[i][j]+  Math.min(up,Math.min(left,right));
            }
            prev=curr;
        }
        
        int result=Integer.MAX_VALUE;
        
        for(int j=0;j<n;j++){
            result=Math.min(result,prev[j]);
        }

        
        return result;
    }
}

File: 931-minimum-falling-path-sum/931-minimum-falling-path-sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
         int n=matrix.size();
        vector<vector<int>> dp(n,vector<int> (n,0));
        for(int i=0;i<n;i++){
            dp[0][i]=matrix[0][i];
        }
        for(int i=1;i<n;i++){
            for(int j=0;j<n;j++){
                int up=INT_MAX,ur=INT_MAX,ul=INT_MAX;
                if(i>0)up=dp[i-1][j];
                if(i>0 && j<n-1)ur=dp[i-1][j+1];
               if(j>0) ul=dp[i-1][j-1];
                dp[i][j]=matrix[i][j]+min(ul,min(up,ur));
            }
        }
        int ans=INT_MAX;
        for(auto i:dp[n-1]){
            ans=min(ans,i);
        }
        return ans;
    }
};